### ISSUE TITLE
UnicodeEncodeError when trying pivot_ui(df)

### ISSUE THREAD

**Issue by Paul-Yuchao-Dong** [nicolaskruchten/jupyter_pivottablejs#2]
Great tool for data exploration, I think this should be a core utility of Jupyter!

I got the error message:
_UnicodeEncodeError: 'ascii' codec can't encode characters in position 8-11: ordinal not in range(128)_

I think it is because of of the pd.to_csv method, adding some encoding handling capability could fix this.

But not really sure how...

thank you for the great js magic!

**Comment by nicolaskruchten** [comment]
OK, I'll see if I can fix this :)

Can you provide a sample data frame that causes this problem?

**Comment by Paul-Yuchao-Dong** [comment]
![image](https://cloud.githubusercontent.com/assets/7532271/9840879/2013f1be-5acb-11e5-9140-01a4902dfbad.png)
Sorry total newbie here, not sure how to do a proper uploading of the data.

Basically, one of the columns contains character in Chinese, like 巴西出口投资促进局北京代表处.

I tested using  pd.to_csv(encoding = 'utf-8'), and it can be properly output into a csv, I am using pandas 0.16+

**Comment by nicolaskruchten** [comment]
I'm having trouble replicating this problem... Can you provide a bit more context/sample code please? What versions of Jupyter/Pandas/Python are you using?

**Comment by ahmetanildindar** [comment]
Same problem :) 
Versions => Jupyter/Pandas/Python = 4.1.0 / 0.18.1 / 2.7.12

![](https://lh3.googleusercontent.com/e5bCW86GwNLHRZmR00mFManqqausjUG5qQJJ1WXDZb3P2limfxpsocJGasU8gp5-Z5W4fiM4WwAyvBqdYVU9PT3kv5FPBmPAFWmQK0A3L5GL6VVjxbwmh5hyAqAU8A8ty7bxX6EpRtpk1u59EfaeLaPd5lf7ivFUwvDK72MOWfIkJPgD0tncrUgg5XsM9-J5aOG80wQw6dZCK-NTgEthUPoT0ps01x59plgdC-BYrlTc2gVaoYZP6PMzkHxl6pLTX8lGsdaumcaMyfVTz8m-jeH3vKH5a1iEuVnWytABd22tUZOSYrOcydinpP42I4oehbNlHEULpmWcKAm0ZYMbe2ciYtl1RJv1CaV-9cvSMXC5u9-NBlB1soaTfpt1qtZjeO3nwiWkn3WTgy6t8kEjhzXwVjZZ-RnUyJudrspSxfURVZo0EmQOMB5w9mBD5CAbM7Ud00kr40QGXwGbnLwSqdZnBAwtOb9IoT9Odhxc83HETCJLFKGlLkWUhcJmsX9FXaAUtixGv4O6LT5fnRGrEKyxhTmTqjkahCwsFXwp0ydmqE_jixlBKWME6ZxAZjvDANt-wylJDOxuviu6je_NQ-I7H8jJiasl=w684-h754-no)
The columns namely "Kullanıcı" and "İsim" cause coding problem as followis
![](https://lh3.googleusercontent.com/-0RE3_cbKjQ5ElptSBFqciFzZ0PcWoIMYMkXMCcEKpnhZRTF8W2x8fBsAPzs3xhIgzh9Nh7KYnZ2CJYeBLNnUOG-OMAB4stuJDcKl45LXemDeOUkhriLg5PxviGAkaolZP9gGCMN9ySpeeWqPqwOPkJQBpcHqGjnkmksjZl0VUgV1FhtjQycBe4uojRKmtyk9XNVPKq935zl7Y-AVMMRa7kiDWiUqErYE1Lmii4ZS5OF0_vgGcVoEzKFJmmYneeQa3mI8Z7NODRW8KcLVgvNCGo3iWE3r_fDztq-1UhOm_1NduWdCvHgJcRPcK42zHxtie0ZPvp0K80wgyEMAugzmJjEG1x-stsOVvXymPcDIxm0uZC9IwEC0-yScyonJlqCBVtdmIDdGzC76OrDGCkzoPN7ykxp8JaU-GcN7uINUgy9PlpGgoduNDbDyj3fwU1i5GyibV0kT3mGIKhP0h8uKbLvIRJjJLncgm3XoBZ2CJ775uekBGmUvU47pDr8e9ZyJ-o0WpYO5sm0WoGZ_47EyE7ed-b8kYwqZFepiyHhF5ezXId_ywR1B_PjUag819Mtq3nTVQCaWszA4e6VNav02QFz87rEMjcu=w830-h754-no)
Any solution or recommendation?

++Ahmet

**Comment by nicolaskruchten** [comment]
Sorry for the long delay on this, but I'm back to giving this project a bit of love... Can someone post a link to a CSV file that exposes this problem? I still can't replicate it.

**Comment by indivisible** [comment]
This reproduces the error:
```
import pandas as pd
import io
from pivottablejs import pivot_ui
csv_data = ',\xe1rv\xedzt\u0171r\u0151 t\xfck\xf6rf\xfar\xf3g\xe9p\n0,42\n'
csv = io.StringIO(csv_data)
df = pd.read_csv(csv)
pivot_ui(df)
```
You should probably treat the input things and template as unicode, and then properly encode it for writing. Easy to fix:
```
diff --git a/pivottablejs/__init__.py b/pivottablejs/__init__.py
index 64c2b2f..fd70da5 100644
--- a/pivottablejs/__init__.py
+++ b/pivottablejs/__init__.py
@@ -1,4 +1,4 @@
-TEMPLATE = """
+TEMPLATE = u"""
 <!DOCTYPE html>
 <html>
     <head>
@@ -70,8 +70,8 @@ import json

 def pivot_ui(df, outfile_path = "pivottablejs.html", url="",
     width="100%", height="500", **kwargs):
-    with open(outfile_path, 'w') as outfile:
-        outfile.write(TEMPLATE %
-            dict(csv=df.to_csv(), kwargs=json.dumps(kwargs)))
+    with open(outfile_path, 'wb') as outfile:
+        outfile.write((TEMPLATE %
+            dict(csv=df.to_csv(), kwargs=json.dumps(kwargs))).encode('utf8'))
     return IFrame(src=url or outfile_path, width=width, height=height)

```

This fixes it for py3.6, and should work for 2.7 too (not tested).

**Comment by nicolaskruchten** [comment]
Interesting. So in Python 2.7, your code results in an error on the line `csv_data = ',\xe1rv\xedzt\u0171r\u0151 t\xfck\xf6rf\xfar\xf3g\xe9p\n0,42\n'` but if I prefix the string with `u` then the code runs without error, including the pivot table:

![image](https://cloud.githubusercontent.com/assets/203523/25918894/f973fd40-359a-11e7-87f3-eb080f317527.png)

**Comment by nicolaskruchten** [comment]
I'm actually seeing the same behaviour in Python 3.5

**Comment by indivisible** [comment]
The problem is that you do things without ever specifying an encoding, so behaviour is system dependent.
See the [docs for example](https://docs.python.org/3/library/functions.html#open):

>  In text mode, if encoding is not specified the encoding used is platform dependent: locale.getpreferredencoding(False) is called to get the current locale encoding.

You should always specify the encoding you wish to use for text IO, even if it happens to work at the time on your dev environment.

**Comment by indivisible** [comment]
Also a cleaner way than my fix would be to just use `io.open(outfile_path, 'w', encoding='utf8')` to open the file.

**Comment by nicolaskruchten** [comment]
Ok I'll look into that, thanks!

To your earlier point, which is the specific call am I making that has
system-dependent behaviour if encoding is not specified?

**Comment by nicolaskruchten** [comment]
Ah never mind, I see the link now :)

**Comment by nicolaskruchten** [comment]
@indivisible your suggestion does work in Python 3 but causes all sorts of encoding errors in Python 2.7 for me with the following CSV: https://github.com/nicolaskruchten/pivottable/blob/master/examples/mps.csv

**Comment by indivisible** [comment]
I see, sorry.

I finally installed python2.7 and looked a bit closer at the problem. I've made a [gist](https://gist.github.com/indivisible/baa7e21e4f6d66d1b123d42fef0ad7ac) that has a script to reproduce the problem, sample outputs, and a slightly better patch.

**Comment by nicolaskruchten** [comment]
OK, nice, this seems like a reasonably elegant fix @indivisible, thanks!

Question re your gist: in `output.txt` the fourth test case fails... Is that with or without your patch? I wasn't able to replicate this test case locally, as here on my Mac even with `LC_ALL=en_US` the preferred locale is still UTF8.

**Comment by indivisible** [comment]
Without. With the patch applied all outputs matched. I have no idea how locale settings on Macs work, the only systems I could test it on were Linux and Windows. The sample output is from Linux. On windows the "preferred encoding" seems to be cp1252...

Cheers!

**Comment by nicolaskruchten** [comment]
Awesome, thanks so much for the help! I've just released v0.7.0 with these fixes :)
